services/galaxytoolshed/galaxy-toolshed-downloads.service.js

```diff
@@ -0,0 +1,55 @@
+import Joi from 'joi'
+import { renderDownloadsBadge } from '../downloads.js'
+import BaseGalaxyToolshedService from './galaxy-toolshed-base.js'
+
+const repositoryRevisionInstallInfoSchema = Joi.array().items(
+  Joi.object({
+    times_downloaded: Joi.number().required(),
+  }).required(),
+  Joi.object({}),
+  Joi.object({})
```

Comment:
As per comments on the base class, this schema definition almost certainly needs to be moved to the base class where the specifics of the associated API are encapsulated.

My impression is that you were trying to use the schemas in more of a class-specific or field-specific manner but that's not really what we want to use the schemas for, and it also adds some unnecessary processing complexity.

Just to reiterate, the schemas define what the response structure should look like for a singular endpoint, and that's still true even in cases like this one where we have multiple badges in a badge family that use different fields from the same response.

What I think we need to do here is to move this schema into the base class to maintain a single definition and encapsulate it alongside its accompanying api call. We can then expand it as needed with additional fields to support additional badges.

Given that the response is structured as a fixed size array with 3 different objects, we need to define our schema to match. At the moment/purely for download count data, that would look something like this:

```suggestion
const repositoryRevisionInstallInfoSchema = Joi.array().ordered(
  Joi.object({
    times_downloaded: Joi.number().required(),
  }).required(),
).items(Joi.any())
```

Note the `.ordered()` call [which guarantees that important ordering](https://joi.dev/api/?v=17.6.0#arrayorderedtype), and the trailing `.items()` which covers the fact that there are other, differently structured elements we don't currently care about. If/when we add more badges that need elements from the second and/or third element from the response, then we can update this schema to match (happy to provide more details when/if we get to that point, but it's basically just adding corresponding entries to the `.ordered()` section

Reply:
Ok, I will update this schema for the other utilities. a6c20a5b3